### DISCUSSION TITLE
ava configuration

### DISCUSSION THREAD

**Discussion by cspotcode** [TypeStrong/ts-node#1247] [Official Recipes] [answered]
This thread is for asking questions about ava & ts-node configuration.

Here is an example of ava and ts-node using import syntax compiled to and executed as CommonJS:
https://github.com/TypeStrong/ts-node-repros/runs/1964972170

Here is an example of ava and ts-node using import syntax compiled to and executed as ESM:
https://github.com/TypeStrong/ts-node-repros/runs/1964966424

**Comment by mesqueeb** [comment]
@cspotcode thank you very much.

Looking at these changes:
 https://github.com/TypeStrong/ts-node-repros/compare/ava-tests-native-esm

For my  `esm` package, I was able able to have make `ts-node` execute my Ava tests only using the `.ts` files.

there are 2 questions:

1. I'm forced to write `import x from '../src/index.js'` even though the file is `index.ts`...
  - This seems bizarre to me and I'm afraid it might break in the future? Is there a special reason we _need_ to write `.js` even though it's a `.ts` file? I read through a lot of documentation you wrote on this in the `esm` thread, but I'm rather confused on this. Would you be open to try to explain again in simple terms for newer devs and also touch on the fact if it might change or not in the future?

2. Every time I execute the tests I get these warnings:
```
 ⚠ Experiments are enabled. These are unsupported and may change or be removed at any time.

(node:64635) ExperimentalWarning: --experimental-loader is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
```
Does this mean that I might need to change some setup again in the future? Is there any threads or news I should follow to know how/what I need to change to still have my Ava tests be executable only with TS?

Thanks a lot!

**Reply by cspotcode** [reply]
> I'm forced to write `import x from '../src/index.js'` even though the file is `index.ts`...

This is enforced by TypeScript, and it is actually done so that things are more stable and *don't* break in the future.  If you think about it, pre-compiling your code means that you are importing `.js` files.  So you want your imports to have the `.js` extension and behave correctly when your code is pre-compiled.  And ts-node's job is to be compatible, so you can run your code in ts-node *or* you can precompile, without any code changes.

Below is a copy of a post I wrote [here](https://github.com/nodejs/modules/issues/351#issuecomment-621257543) to explain why you must `import from './foo.js';` to import `foo.ts`:

`foo.js` resolves to `foo.js`, which is the compiled output of `foo.ts` and does not exist on disk prior to compilation.  `ts-node` runs prior to compilation.

TypeScript does not modify valid ES syntax unless explicitly asked to.  JSX is transformed because it is not valid ES.  Type annotations are removed because they are not valid ES.  `import from './foo.js';` is emitted verbatim, because it *is* valid ES syntax.  (it will be transformed if we ask TS to convert to CommonJS or downlevel to ES3)  At design time, the language service understands the semantics, because it knows `./foo.js` is the emitted code corresponding to `./foo.ts`, `./foo.tsx`, or `./foo.jsx`.

This gives developers explicit control over their import module specifiers at runtime, and those specifiers can remain stable for certain changes to the codebase.  For example, renaming a `.js` file to `.jsx` does not requiring changing the import statement.  Refactoring `.js` into `.ts`, or vice versa, is the same.

`ts-node`'s goal is to be a drop-in replacement for the compile step.  If you would normally `tsc && node ./index`, we allow you to `ts-node ./index`.  We also need to maintain compatibility with the larger ecosystem, in particular the language service, because that is the value of TS.

> Every time I execute the tests I get these warnings:

Where are those warnings coming from?  Node, ava, or both?

Remember that node's loader hooks are experimental.  https://nodejs.org/dist/latest-v15.x/docs/api/esm.html#esm_loaders

**Reply by cspotcode** [reply]
Also, in #1007 we mention that we are compatible with node's `--experimental-specifier-resolution=node`.  Check out node's documentation here: https://nodejs.org/dist/latest-v15.x/docs/api/esm.html#esm_customizing_esm_specifier_resolution_algorithm

**Comment by cspotcode** [comment, marked as answer]
If the recipe above can be improved in any way, please ask questions here and we will update it accordingly.
